feat: Add facility function tests and fix missing test targets

rs0125 opened the pull request:
### Improve Facility Error Handling (Fixes https://github.com/goatshriek/stumpless/issues/456)

This PR improves error handling and test coverage for the facility-related functions in the library. It also introduces a new function test target to aid future debugging.

---

###  Code Changes

#### `src/facility.c`

* **`stumpless_get_facility_string`**:

  * Now calls `clear_error()` on success.
  * Calls `raise_error(STUMPLESS_INVALID_FACILITY)` for invalid enums.

* **`stumpless_get_facility_enum_from_buffer`**:

  * Calls `clear_error()` on success.
  * Calls `raise_error(STUMPLESS_INVALID_FACILITY)` for unmatched strings.

* **`stumpless_get_facility_enum`**:

  * Now clears error if the call to `*_from_buffer` was successful.

---

###  Test Additions (`test/function/facility.cpp`)

Added tests in a new **`FacilityErrorHandling`** suite:

* `ClearsPreviousErrorOnSuccess`: ensures previous errors are cleared.
* `InvalidFacilityEnumRaisesError`: checks invalid enum detection.
* `InvalidFacilityStringRaisesError`: checks invalid string input.

---

###  Function Target

* Added a `function-test-facility` CMake target to run facility tests independently (mirroring `function-test-version` pattern).
* All 8 tests pass successfully.

![image](https://github.com/user-attachments/assets/426dfd3d-c862-4941-8a7d-1c4ad9601252)


---

###  Doxygen Header Updates (`include/stumpless/facility.h`)

Each relevant function comment now includes:

* `@error This function clears any previous error on success.`
* `@error This function sets STUMPLESS_INVALID_FACILITY on invalid input.`

---

rs0125 committed:
feat: Add facility function tests and fix missing test targets

goatshriek commented:
Thanks very much for putting this together, particularly for describing your updates so thoroughly in the PR!

As you work through these errors (the header check static analysis should help), please also remove the formatting changes from this PR. The changes should be limited to just the changes you are adding, and not reformatting if other parts of updated files.

rs0125 committed:
Revert "feat: Add facility function tests and fix missing test targets"

This reverts commit 64e0951dce63141966f7010e113d38c3356ddc44.

rs0125 committed:
feat: Add facility function tests and fix missing test targets (without white space changes and fixed header)

rs0125 commented:
@goatshriek would it be possible to run a check now? 


There was a missing header somewhere which was causing the issues (private/error.h) 
Added that and that should have fixed it. 
Also cleaned up all the reformatting changes.

Also is there any way I can run the CI tests on my own (like during some other contribution) before sending the commit?

goatshriek commented:
> Also is there any way I can run the CI tests on my own (like during some other contribution) before sending the commit?

There are two straightforward ways to do this, the first and probably easiest being to enable GitHub Actions on your fork. That should allow you to recreate the exact checks being run for the PR without waiting for a maintainer to approve the runs. You can also go through the workflow files and run the tests manually yourself in your dev environment, but this requires you to install dependencies and tooling, and is a bit more work and more error prone.

rs0125 committed:
fix: Added stumpless/error.h header file

rs0125 commented:
Had another header file issue

This should fix it

goatshriek commented on include/stumpless/facility.h:
```diff
@@ -389,6 +389,9 @@ stumpless_get_facility_string( enum stumpless_facility facility );
  *
  * @param facility_string The facility name to get the enum from.
+ * 
+ * @error Clears previous error on success.
+ * @error Raises STUMPLESS_INVALID_FACILITY on invalid input.
```
As far as I can tell from the [Doxygen documentation](https://www.doxygen.nl/manual/commands.html) `@error` is not a supported special command. Instead, you can just put these two descriptions into their own paragraph to describe the error behaviors.

goatshriek commented on include/stumpless/facility.h:
```diff
@@ -415,6 +418,9 @@ stumpless_get_facility_enum( const char *facility_string );
  * 
  * @param facility_buffer_length The length of the buffer
+ * 
+ * @error Clears previous error on success.
+ * @error Raises STUMPLESS_INVALID_FACILITY on invalid input.
```
Same here as above, replace the non-existent `@error` commands with a description of the behavior.

goatshriek commented on src/facility.c:
```diff
@@ -29,44 +31,58 @@ static char *facility_enum_to_string[] = {
+  return result;
 }
 
+
+
```
Remove these extra newlines that were added.

goatshriek commented on test/function/facility.cpp:
```diff
@@ -116,4 +116,26 @@ namespace {
     EXPECT_EQ( result, -1 );
   }
 
+  TEST( FacilityErrorHandling, ClearsPreviousErrorOnSuccess ) {
```
Please indent the body of each new test function.

goatshriek reviewed (changes requested):
Overall this looks good! Just a few formatting changes for you to make, and this should be good to go once the headers are also worked out.

rs0125 committed:
fix: fixed header and formatting

rs0125 commented:
That should be all. Please let me know if there's any other changes
Thanks a lot for your patience

rs0125 commented:
I wonder why the test scripts need reimports of headers as different instances when they're already calling scripts from the base facility.c file. The CI workflow keeps flagging it even though it runs on compilation. Is there any particular reason for having that check?

Anyway I've ammended my last commit. Pardon me for making this go on for longer than it should have.

goatshriek commented:
Thanks again for sticking with this through the change requests!

The tests are checked separately for includes since they are different compilation units, and likely even use a different compiler than the .c files in the library itself (gcc vs. g++, for example). It's important to make sure that everything used in a single *.cpp or *.c file is included in it, so that there is not an implicit dependency on one header including another, which could lead to unexpected breakage. [IWYU](https://include-what-you-use.org/) is a standard tool for enforcing this, and has more information on the philosophy, if you'd like to learn more about it. We use a custom script here for the checks, to reduce the number of dependencies used for dev, since Ruby is already used for C++ binding generation.

rs0125 commented:
Thanks a lot for the thorough reviews and guidance throughout! Looking forward to contributing more to this project in the future. 

Really appreciate the time and feedback
